src/zenml/integrations/aws/step_operators/sagemaker_step_operator.py

```diff
@@ -0,0 +1,145 @@
+#  Copyright (c) ZenML GmbH 2022. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+
+from typing import List, Optional
+
+import sagemaker
+
+from zenml.enums import StackComponentType, StepOperatorFlavor
+from zenml.repository import Repository
+from zenml.stack import StackValidator
+from zenml.stack.stack_component_class_registry import (
+    register_stack_component_class,
+)
+from zenml.step_operators import BaseStepOperator
+from zenml.utils import docker_utils
+
+
+@register_stack_component_class(
+    component_type=StackComponentType.STEP_OPERATOR,
+    component_flavor=StepOperatorFlavor.SAGEMAKER,
+)
+class SagemakerStepOperator(BaseStepOperator):
+    """Step operator to run a step on Sagemaker.
+
+    This class defines code that buidls an image with the ZenML entrypoint 
+    to run using Sagemaker's Estimator.
+
+    Attributes:
+        role: The role that has to be assigned to jobs running in Sagemaker.
+        instance_type: The instance type of the compute where jobs will run. 
+        base_image: [Optional] The base image to use for building the docker 
+            image that will be executed.  
+        bucket: [Optional] Name of the S3 bucket to use for storing artifacts 
+            from the job run. If not provided, a default bucket will be created 
+            based on the following format: "sagemaker-{region}-{aws-account-id}".
+        experiment_name: [Optional] The name for the experiment to which the job 
+            will be associated. If not provided, the job runs would be independent. 
+    """
+
+    supports_local_execution = True
+    supports_remote_execution = True
+
+    role: str
+    instance_type: str
+
+    base_image: Optional[str] = None
+    bucket: Optional[str] = None
+    experiment_name: Optional[str] = None
+
+    @property
+    def flavor(self) -> StepOperatorFlavor:
+        """The step operator flavor."""
+        return StepOperatorFlavor.SAGEMAKER
+
+    @property
+    def validator(self) -> Optional[StackValidator]:
+        """Validates that the stack contains a container registry."""
+        return StackValidator(
+            required_components={StackComponentType.CONTAINER_REGISTRY}
+        )
```

Comment:
What is this for? I see that we do a manual check for the container_registry inside `_build_docker_image` below, but we don't use this `validator`.

Reply:
This check happens really early before the pipeline itself is orchestrated. Later on during the step execution the check is mostly for mypy to verify that the container registry is not None, but logically the other error should never occur